Firestore/Protos/nanopb_pretty_printing/generators.py

```diff
@@ -0,0 +1,415 @@
+    """
+
+    count = self.name + '_count'
+
+    result = _reindent(1, '''\
+        for (pb_size_t i = 0; i != %s; ++i) {\n''') % count
+    # If the repeated field is non-empty, print all its members, even if they are
+    # zero or empty (otherwise, an array of zeroes would be indistinguishable from
+    # an empty array).
+    result += self._generate_for_leaf(indent=2, always_print=True)
+    result += _reindent(1, '''\
+        }\n''')
+
+    return result
+
+  def _generate_for_optional(self):
+    """Generates a C++ statement that prints the optional field, if set.
+    """
+
+    name = self.name
+    result = _reindent(1, '''\
+        if (has_%s) {\n''') % name
+    # If an optional field is set, always print the value, even if it's zero or
+    # empty.
+    result += self._generate_for_leaf(indent=2, always_print=True)
+    result += _reindent(1, '''\
+        }\n''')
+
+    return result
+
+  def _generate_for_leaf(self, indent=1, always_print=False, parent_oneof=None):
+    """Generates a C++ statement that prints the "leaf" field.
+
+    "Leaf" is to indicate that this function is non-recursive. If the field is
+    a message type, the generated code will delegate printing to its
+    `ToString()` member function.
+
+    Args:
+      indent: The indentation level of the generated statement.
+      always_print: If `False`, the field will not be printed if it has the
+        default value, or for a message, if each field it contains has the
+        default value.
+      parent_oneof: If the field is a member of a oneof, a reference to the
+        corresponding `OneOfPrettyPrintingGenerator`
+    """
+
+    always_print = 'true' if always_print else 'false'
+
+    display_name = self.name
+    if self.is_primitive:
+      display_name += ':'
+
+    cc_name = self._get_cc_name(parent_oneof)
+    function_name = self._get_printer_function_name()
+
+    return self._generate(indent, display_name, cc_name, function_name,
+                          always_print)
+
+  def _get_cc_name(self, parent_oneof):
+    """Gets the name of the field to use in the generated C++ code:
+
+    - for repeated fields, appends indexing in the form of `[i]`;
+    - for named union members, prepends the name of the enclosing union;
+    - ensures the name isn't a C++ keyword by appending an underscore
+      (currently, only for keyword `delete`).
+    """
+
+    cc_name = self.name
+
+    # If a proto field is named `delete`, it is renamed to `delete_` by our script
+    # because `delete` is a keyword in C++. Unfortunately, the renaming mechanism
+    # runs separately from generating pretty printers; consequently, when pretty
+    # printers are being generated, all proto fields still have their unmodified
+    # names.
+    if cc_name == 'delete':
+      cc_name = 'delete_'
+
+    if self.is_repeated:
+      cc_name += '[i]'
+
+    if parent_oneof and not parent_oneof._is_anonymous:
```

Comment:
You've accessing `_is_anonymous` from outside the class that defines it. This means that it shouldn't be a "private" member.